### PR TITLE
[Gekidou] Solidarity poc

### PR DESCRIPTION
**Pull request by ElonVolo**
#### Summary

Adds a npm pre-install script that uses the [Solidarity](https://infinitered.github.io/solidarity/#/) dependency checker to ensure that all the global dependencies (Xcode, npm, react-native-cli, environment variables, etc) required to build mattermost-mobile are installed on a developer's machine. If all the required dependencies are found to be present on a developer's machine, the preinstall script passes and the npm install is allowed to proceed; if some dependencies are missing or are an incorrect version, then the npm install for mattermost-mobile is halted until the necessary dependencies are installed and correctly configured.

#### Ticket Link

This is a possible fix for issues such as https://github.com/mattermost/mattermost-mobile/issues/6329. 

If the user is can be informed during the dependency install process that they don't have the necessary global dependencies required to build the app, considerable amounts of developer time on both ends can be saved as it avoids back-and-forths on newly reported github issues tickets that need to be solved by telling the reporter that they need to up/downgrade a different version of a dependency.

#### Device Information
This PR was tested on: 

#### Screenshots

Screen Shot 2022-06-10 at 3 07 43 AM

#### Release Note

```release-note
Added a npm pre-install script that uses Solidarity to check for required build dependencies.
```

### CODE REVIEW

**Review by enahum — CHANGES_REQUESTED** [review]
Thanks for this @ElonVolo a quick review with a comment

File: .solidarity [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      ],
+      "Package.json": [{ "rule": "file", "location": "./package.json"}],
+      "React Native": [
+        { "rule": "cli", "binary": "react-native", "semver": "~2.0.1"},
+        { "rule": "cli", "binary": "react-native", "semver": ">0.0.1", "line": 2 }
+      ],
+      "Xcode" : [
+        { "rule": "cli", "binary": "xcodebuild", "semver": ">=11.0", "platform": "darwin"}
+      ],
+      "Cocoapods": [
```
[the remark this answers is not in the document]
Comment (ElonVolo) [reply .solidarity:36]: I have the two following questions.

1. Should take out the CocoaPods check in or leave it in there?

One the of advantages of putting all the checks in a dependence checking is that the dependency checker file (in this case the .solidarity file) effectively turns into a single source of truth document that lists everything required to do the build and that can't become outdated like a README.md or a wiki page; if the config file's outdated nothing works and people know immediately it must be updated (to get stuff to work again). But if that would hamper some workflow everyone is using I can remove the .solidarity config file entry for Cocoapods.

2. Is exactly version 2.0.2 of Bundler (mentioned in the Mattermost dev environment set up docs) required ? I'm currently on Ruby v3.1.2 (via rvm) and when I do ```gem install bundler --version 2.0.2```, I end up with Bundler version 2.3.7 in spite of my specifying the earlier version (who knows?). Is there actually a range of Bundler versions I should be targeting instead?

**Comment by maxgorovenko** [comment]
This check could save me couple hours :(

**Review by enahum — APPROVED** [review]

**Comment by enahum** [comment]
@ElonVolo thanks for bringing this to our attention and for laying the ground work for this.

**Comment by ElonVolo** [comment]
No prob. 😀 I'm always happy to promote my philosophy of pro-active
synchronization and standardization of global dependencies across developer
machines, as I feel that people should spend their time actually developing
valuable and useful features for software instead of spending frustrating
days of trying to figure out why their development environment isn't
building stuff (when their coworker's strangely seems to have no issue).
I've found this approach is one of the cheapest and easiest things you can
do on a project to effectively and immediately boost developer velocity.

**Pull request merged**